resize-observer/devicepixel2.html

```diff
@@ -0,0 +1,163 @@
+<!doctype html>
+<link rel="match" href="devicepixel2-ref.html">
+<meta name="assert" content="Device pixel content box sizes and pixel snapping are correct in Resize Observer callback">
+<style>
+  html, body {
+    margin: 0;
+  }
+  #front {
+    position: absolute;
+    left: 0;
+    top: 0;
+  }
+  .outer {
+    width: 100px;
+    display: flex;
+  }
+  .outer>* {
+    flex: 1;
+    height: 20px;
+  }
+</style>
+<body>
+  <div id="front"></div>
+</body>
+<script>
+const patternSize = 4;
+
+function createPatternDataURL() {
+  const ctx = document.createElement('canvas').getContext('2d');
+  ctx.canvas.width = patternSize;
+  ctx.canvas.height = patternSize;
+
+  const b = [0, 0, 0, 255];
+  const t = [0, 0, 0, 0];
+  const r = [255, 0, 0, 255];
+  const g = [0, 255, 0, 255];
+
+  const imageData = new ImageData(patternSize, patternSize);
+  imageData.data.set([
+    b, t, t, r,
+    t, b, g, t,
+    t, r, b, t,
+    g, t, t, b,
+  ].flat());
+  ctx.putImageData(imageData, 0, 0);
+  return ctx.canvas.toDataURL();
+}
+
+const patternURL = createPatternDataURL();
+
+/**
+ * Set the pattern's size on this element so that it draws were
+ * 1 pixel in the pattern maps to 1 devicePixel and then set
+ * its position so it's aligned to the pattern of the body element.
+ * The pattern is 'patternSize' large. x and y are the computed 
+ * devicePixel position of the element. To make the pattern align
+ * we need to set its position to -(x % patternSize), -(y % patternSize)
+ * in devicePixels
+*/
+function setPattern(elem, x, y) {
+  const oneDevicePixel = 1 / devicePixelRatio;
+  const patternPixels = oneDevicePixel * patternSize;
+  const xOff = -(x % patternSize);
+  const yOff = -(y % patternSize);
```

Comment:
(optional): I think you could set the background-position (xOff and yOff) to just `-x -y` in device pixels (no modulus operation needed), right?

Essentially, we'd be setting every element's background-position to 0 0 in the "global coordinate system (which for an element at `x,y` is `-x, -y`).  Then it will be trivially aligned/properly-overlaid, as long as we have the correct device-pixel positioning (i.e. as long as we don't have the Firefox bug).

The current modulus setup works too, but feels slightly more complex / less intuitive, IMO.

Reply:
done